src/raftstore/store/peer.rs

```diff
@@ -802,6 +802,78 @@ impl Peer {
         cmd.call(resp);
     }
 
+    /// Calculate the number of the healthy nodes.
+    /// A node is healthy when
+    /// 1. it's the leader of the Raft group, which has the latest logs
+    /// 2. it's a follower, and it does not lag behind the leader a lot.
+    ///    If a snapshot is involved between it and the Raft leader, it's not healthy since
+    ///    it cannot works as a node in the quorum to receive replicating logs from leader.
+    fn calc_the_healthy(&self, progress: &HashMap<u64, Progress>) -> usize {
+        let mut healthy = 0;
+        for pr in progress.values() {
+            if pr.matched >= self.get_store().truncated_index() {
+                healthy += 1;
+            }
+        }
+        healthy
+    }
+
+    /// Check whether it's safe to propose the specified conf change request.
+    /// It's safe iff at least the quorum of the Raft group is still healthy
+    /// right after that conf change is applied.
+    /// Define the total number of nodes in current Raft cluster to be `total`.
+    /// To ensure the above safety, if the cmd is
+    /// 1. A `AddNode` request
+    ///    Then at least '(total + 1)/2 + 1' nodes need to be up to date for now.
+    /// 2. A `RemoveNode` request
+    ///    Then at least '(total - 1)/2 + 1' other nodes (the node about to be removed is excluded)
+    ///    need to be up to date for now.
+    fn check_conf_change(&self, cmd: &RaftCmdRequest) -> Result<()> {
+        let change_peer = get_change_peer_cmd(cmd).unwrap();
+
+        let change_type = change_peer.get_change_type();
+        let peer = change_peer.get_peer();
+
+        let mut status = self.raft_group.status();
+        let total = status.progress.len();
+        if total == 1 {
+            // It's always safe if there is only one node in the cluster.
+            return Ok(());
+        }
+
+        match change_type {
+            ConfChangeType::AddNode => {
+                let progress = Progress { ..Default::default() };
+                status.progress.insert(peer.get_id(), progress);
+            }
+            ConfChangeType::RemoveNode => {
+                if status.progress.remove(&peer.get_id()).is_none() {
+                    // It's always safe to remove a unexisting node.
+                    return Ok(());
+                }
+            }
+        }
+        let healthy = self.calc_the_healthy(&status.progress);
+        let quorum_after_change = raft::calc_quorum(status.progress.len());
+        if healthy >= quorum_after_change {
+            return Ok(());
+        }
+
+        info!("{} rejects unsafe conf change request {:?}, total {}, healthy {},  \
```

Comment:
please add metric here.
If the reject number is too high, something may be wrong.